cert_test.go

```diff
@@ -0,0 +1,770 @@
+package main
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/tls"
+	_ "crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"io/ioutil"
+	"math/big"
+	"net"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/TykTechnologies/tyk/apidef"
+	"github.com/TykTechnologies/tyk/certs"
+	"github.com/TykTechnologies/tyk/config"
+)
+
+func getTLSClient(cert *tls.Certificate, caCert []byte) *http.Client {
+	// Setup HTTPS client
+	tlsConfig := &tls.Config{}
+
+	if cert != nil {
+		tlsConfig.Certificates = []tls.Certificate{*cert}
+	}
+
+	if len(caCert) > 0 {
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+		tlsConfig.RootCAs = caCertPool
+		tlsConfig.BuildNameToCertificate()
+	} else {
+		tlsConfig.InsecureSkipVerify = true
+	}
+
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+
+	return &http.Client{Transport: transport}
+}
+
+func genCertificate(template *x509.Certificate) ([]byte, []byte, []byte, tls.Certificate) {
+	priv, _ := rsa.GenerateKey(rand.Reader, 2048)
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, _ := rand.Int(rand.Reader, serialNumberLimit)
+	template.SerialNumber = serialNumber
+	template.BasicConstraintsValid = true
+	template.NotBefore = time.Now()
+	template.NotAfter = time.Now().Add(time.Hour)
+
+	derBytes, _ := x509.CreateCertificate(rand.Reader, template, template, &priv.PublicKey, priv)
+
+	var certPem, keyPem bytes.Buffer
+	pem.Encode(&certPem, &pem.Block{Type: "CERTIFICATE", Bytes: derBytes})
+	pem.Encode(&keyPem, &pem.Block{Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(priv)})
+
+	clientCert, _ := tls.X509KeyPair(certPem.Bytes(), keyPem.Bytes())
+
+	combinedPEM := bytes.Join([][]byte{certPem.Bytes(), keyPem.Bytes()}, []byte("\n"))
+
+	return certPem.Bytes(), keyPem.Bytes(), combinedPEM, clientCert
+}
+
+func genServerCertificate() ([]byte, []byte, []byte, tls.Certificate) {
+	certPem, privPem, combinedPEM, cert := genCertificate(&x509.Certificate{
+		DNSNames:    []string{"localhost"},
+		IPAddresses: []net.IP{net.ParseIP("127.0.0.1"), net.ParseIP("::")},
+	})
+
+	return certPem, privPem, combinedPEM, cert
+}
+
+func TestGatewayTLS(t *testing.T) {
+	// Configure server
+	serverCertPem, serverPrivPem, combinedPEM, _ := genServerCertificate()
+
+	config.Global.HttpServerOptions.UseSSL = true
+	config.Global.ListenPort = 0
+
+	dir, _ := ioutil.TempDir("", "certs")
+
+	defer func() {
+		os.RemoveAll(dir)
+		config.Global.HttpServerOptions.UseSSL = false
+		config.Global.ListenPort = 8080
+	}()
+
+	t.Run("Without certificates", func(t *testing.T) {
```

Comment:
I would avoid spaces in the names, and use camelcase instead like `WithoutCerts` to be consistent with the parent test names

Reply:
I would prefer staying like this because this is one of the features I like with using `t.Run`. For me, it is the replacement of comments in the code, so I can in human language describe the high-level goal of the test. And in the end, it works fine when you run the single nested test from the command line too, it just replaces spaces with underlines.

Reply:
We agree to disagree :) I simply don't like the name `CamelCase/With_many_underscores`, it seems unnecessary to mix the two styles.